python-sdk/src/uto/data_providers/aws/s3.py

```diff
@@ -0,0 +1,149 @@
+from __future__ import annotations
+
+import logging
+import os
+
+from airflow.hooks.dbapi import DbApiHook
+from airflow.providers.amazon.aws.hooks.s3 import S3Hook
+from uto.data_providers.base import DataProviders
+from uto.datasets.base import UniversalDataset as Dataset
+from uto.utils import FileLocation, create_dataprovider, get_dataset_connection_type
+
+from astro.constants import LoadExistStrategy
+
+
+class S3DataProvider(DataProviders):
+    """
+    DataProviders interactions with S3 Dataset.
+    """
+
+    def __init__(
+        self,
+        conn_id: str,
+        optimization_params: dict,
+        extras: dict = {},
+        use_optimized_transfer: bool = True,
+        if_exists: LoadExistStrategy = "replace",
+    ):
+        super().__init__(
+            conn_id=conn_id,
+            extras=extras,
+            optimization_params=optimization_params,
+            use_optimized_transfer=use_optimized_transfer,
+            if_exists=if_exists,
+        )
+        self.transfer_mapping: set = {
+            FileLocation.GS,
+            FileLocation.S3,
+            FileLocation.AWS,
+            FileLocation.google_cloud_platform,
+        }
+
+    @property
+    def hook(self) -> DbApiHook:
+        """Return an instance of the database-specific Airflow hook."""
+        return S3Hook(
+            aws_conn_id=self.conn_id,
+            verify=self.extras.get("verify", None),
+            transfer_config_args=self.extras.get("transfer_config_args", None),
+            extra_args=self.extras.get("s3_extra_args", {}),
+        )
+
+    def check_if_exists(self, dataset: Dataset) -> bool:
+        """Return true if the dataset exists"""
+        raise NotImplementedError
+
+    def check_if_transfer_supported(self, source_dataset: Dataset) -> bool:
+        """
+        Checks if the transfer is supported from source to destination based on source_dataset.
+        """
+        source_connection_type = get_dataset_connection_type(source_dataset)
+        return FileLocation(source_connection_type) in self.transfer_mapping
+
+    def load_data_from_source(self, source_dataset: Dataset, destination_dataset: Dataset) -> None:
+        """
+        Loads data from source dataset to the destination using data provider
+        """
+        if not self.check_if_transfer_supported(source_dataset=source_dataset):
+            raise ValueError("Transfer not supported yet.")
+        source_connection_type = get_dataset_connection_type(source_dataset)
+        if source_connection_type == "google_cloud_platform":
```

Comment:
if we integrate more sources this if else block will become a bit messy

Reply:
Yes changed it to dynamically load from source to destination as part of this commit: https://github.com/astronomer/astro-sdk/pull/1267/commits/fa8bf608338638f5db34bc01edb35cf027ac3740